src/lib/reasoners/theory.ml

```diff
@@ -552,24 +493,31 @@ module Main_Default : S = struct
             Printer.print_dbg
               "bottom (case-split):%a"
               Expr.print_tagged_classes classes;
-          let objectives = partial_objectives_reset env.objectives order in
-          let env = { env with objectives } in
-          propagate_choices env sem_facts acc_choices
+          let env =
+            match order with
+            | Some i ->
+              { env with
+                objectives = Objective.Model.reset_until env.objectives i }
+            | None -> env
+          in
+          propagate_choices env sem_facts acc_choices acc_objs
             [neg_c, lit_orig, CNeg, dep, order]
 
-    and aux env sem_facts acc_choices new_choices =
+    and aux env sem_facts acc_choices (objs, order) new_choices =
       Options.tool_req 3 "TR-CCX-CS-Case-Split";
       match new_choices with
       | [] ->
-        begin match next_optimization ~for_model env with
-          | Some opt_split ->
-            optimizing_split env sem_facts acc_choices opt_split
-          | None -> generate_choices env sem_facts acc_choices
+        begin match objs with
+          | obj :: objs ->
+            optimizing_objective env sem_facts acc_choices (objs, order) obj
+          | [] ->
+            generate_choices env sem_facts acc_choices
         end
       | _ ->
-        propagate_choices env sem_facts acc_choices new_choices
+        propagate_choices env sem_facts acc_choices (objs, order) new_choices
     in
-    aux env sem_facts (List.rev env.choices) new_choices
+    aux env sem_facts (List.rev env.choices)
+      (Objective.Model.functions env.objectives, 0) new_choices
```

Comment:
Is starting at order `0` here really correct? What if some objectives have been optimized already?

Reply:
Hm actually it looks like we are optimizing all objectives again at each case split, that looks wasteful (possibly incorrect, but I don't think so since we use equalities for the splits).

Reply:
Let's clarify a bit the previous behaviour and the current one.
- Before this PR, the function `look_for_sat` proceeds choices in this order when `for_model = false`:
  + First, it propagates new choices in the pending queue.
  + Then it try to optimize as much as it can objectives for which we didn't get a best value yet (in the current SAT branch of course). If it see an objective whose the previous value is a limit (+oo, -oo or a strict bound), it tries to optimize it again.
  + If we fail to optimize a bound, we stop in `optimizing_objective`.
  + Otherwise, it asks theories to generate new case splits and it goes back to the first point.

Now, if we're generating model (so `for_model = true`), the previous code ignore objectives whose the value is a limit and we keep optimize next objectives (with a lower priority).

This is not what we want. If we got a limit for some objective `o`, it doesn't make sense to try to optimize objectives with lower priority than `o`. But in model generation, we want to keep generating model, even if we can fulfill some objectives.

Let's pick an example:
``` smt
(set-option :produce-models true)
(set-logic ALL)
(declare-const x Int)
(declare-const y Int)
(maximize x)
(minimize y)
(assert (>= y (- 10)))
(assert (>= x 0))
(check-sat)
(get-model)
(assert (<= x 10))
(check-sat)
(get-model)
```
If you run AE `next` on it with the command
```sh
dune exec -- alt-ergo test.smt2 --optimize --objctives-in-interpretation --frontend legacy --dump-models
```
AE outputs:
```
unknown
Returned unknown reason = Incomplete
(
  (define-fun x () Int +oo)
  (define-fun y () Int (- 10))
)

unknown
Returned unknown reason = Incomplete
(
  (define-fun x () Int 10)
  (define-fun y () Int (- 10))
)
```
We don't expect that AE keeps optimize value after `x` after the first `check-sat`, but we keep doing it... 

Now, if we don't the same on this PR with the command
```sh
dune exec -- alt-ergo test2.smt2 --optimize --objectives-in-interpretation
```
AE outputs:
```
unknown
(
  (define-fun x () Int 1)
  (define-fun y () Int 0)
)

; File "test2.smt2", line 12, characters 1-12: I don't know (0.0009) (6 steps) (goal g_2)

unknown
(
  (define-fun x () Int 10)
  (define-fun y () Int (- 10))
)
```

The new code does the right thing... by chance! It's because during each case split phase, we do again all the optimizations, even if we got an infinity value before and we stop optimization if we got an infinity value in `optimizing_objective`. If the flag `for_model` is `true`, we stop also optimization but we keep generate new casesplits. 

I add a new function `next_unknown` in `Objective.Model` whose the semantic is closed to the old `next_optimization`:
```ocaml
  exception Found of Function.t * int

  let next_unknown mdl =
    try
      M.iter (fun { f; order } v ->
        match (v : Value.t) with
        | Value _ -> ()
        | Pinfinity | Minfinity | Limit _ -> raise Exit
        | _ -> raise (Found (f, order))
      ) mdl;
      None
    with
    | Found (f, order) -> Some (f, order)
    | Exit -> None
```

Notice this function isn't very efficient because we keep visiting `Value` node in the map again and again. The behaviour of `Objective.Model` is basically a queue with some printers. I guess I will rewrite it with a functional queue.

Reply:
Ok, it's still incorrect for this test:
```smt
; This test checks if the optimization doesn't stop after trying to
; optimize a strict bound.
(set-option :produce-models true)
(set-logic ALL)
(declare-const x Real)
(declare-const y Real)
; If the SAT solver explores first the right branch of the following OR gate, it
; shouldn't conclude that x cannot be optimized.
(assert (or (<= x 4.0) (< x 1.5)))
(maximize x)
; The current implementation of SatML explores the right branch of OR gates
; first. We add the following line to be sure the optimization feature still
; works after changing the order of branches.
(assert (or (< y 1.5) (<= y 4.0)))
(maximize y)
(check-sat)
(get-model)
(get-objectives)
``` 
During the first assert we don't know that `y` will be bound and we get `infinity` here. Thus, we have to treat infinity value again after the second assert, otherwise we will print `+oo` for `y`. This was explained in a comment of the original OptimAE PR.

Reply:
> We don't expect that AE keeps optimize value after `x` after the first `check-sat`, but we keep doing it...

I disagree: in that case I fully expect alt-ergo to re-optimize the model because of the new `get-sat` and the new constraints added by `assert` since the last `check-sat` !

Reply:
Ok my example is ambiguous. I was talking of the first `check-sat` only. I totally agree we have to keep optimize if new constraints are involved.